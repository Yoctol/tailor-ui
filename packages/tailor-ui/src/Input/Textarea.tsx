import React, {
  ChangeEvent,
  FunctionComponent,
  TextareaHTMLAttributes,
  forwardRef,
} from 'react';
import TextareaAutosize from 'react-autosize-textarea';
import styled from 'styled-components';
import { omit } from 'ramda';

import { mergeEventProps } from '@tailor-ui/utils';

import useFormField from '../FormField/useFormField';

import { InputProps } from './Input';
import { inputStyles } from './styles';

export type TextareaProps = InputProps &
  TextareaHTMLAttributes<any> & {
    resize?: boolean;
  };

export const StyledTextarea = styled(props => (
  <TextareaAutosize {...omit(['invalid'], props)} />
))<TextareaProps>`
  word-wrap: break-word;
  resize: ${({ resize }) => (resize ? 'initial' : 'none')};

  ${inputStyles};

  /* stylelint-disable-next-line order/properties-order */
  height: auto;
  transition: border 0.2s ease;
`;

const Textarea: FunctionComponent<TextareaProps> = forwardRef<
  HTMLInputElement,
  InputProps
>(({ id, ...props }, ref) => {
  const [invalid, labelId, setValue] = useFormField({
    id,
    value: props.value,
    defaultValue: props.defaultValue,
  });

  return (
    <StyledTextarea
      ref={ref}
      id={labelId}
      invalid={invalid || undefined}
      {...mergeEventProps(props, {
        onChange: (event: ChangeEvent<HTMLTextAreaElement>) => {
          setValue(event.currentTarget.value);
        },
      })}
    />
  );
});

export default Textarea;
