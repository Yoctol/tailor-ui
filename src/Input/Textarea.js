import PropTypes from 'prop-types';
import React from 'react';
import TextareaAutosize from 'react-textarea-autosize';
import styled from 'styled-components';

import { StyledInput } from '.';

export const StyledTextarea = styled(
  StyledInput.withComponent(TextareaAutosize)
)`
  word-wrap: break-word;
  transition: border 0.2s ease;
  resize: ${({ resize }) => (resize ? 'initial' : 'none')};
`;

const Textarea = props => <StyledTextarea size="md" {...props} />;

Textarea.propTypes = {
  /**
   * Specific the max rows of textarea
   */
  maxRows: PropTypes.number,
  /**
   * Specific the min rows of textarea
   */
  minRows: PropTypes.number,
  /**
   * The callback function that is triggered when height changed
   */
  onHeightChange: PropTypes.func,
};

Textarea.defaultProps = {
  maxRows: null,
  minRows: null,
  onHeightChange: () => {},
};

export default Textarea;
