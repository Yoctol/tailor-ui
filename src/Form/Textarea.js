import styled from 'styled-components';
import TextareaAutosize from 'react-textarea-autosize';
import { ifProp } from 'styled-tools';

import Input from './Input';

const Textarea = styled(Input.withComponent(TextareaAutosize))`
  word-wrap: break-word;
  transition: border 0.2s ease;
  resize: ${ifProp('resize', 'initial', 'none')};
`;

export default Textarea;
