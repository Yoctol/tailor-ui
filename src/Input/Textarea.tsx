import React, { FunctionComponent } from 'react';
import TextareaAutosize from 'react-textarea-autosize';
import styled from 'styled-components';

import { InputProps, inputStyles } from './Input';

export type TextareaProps = InputProps & {
  resize?: boolean;
};

export const StyledTextarea = styled(TextareaAutosize)<TextareaProps>`
  word-wrap: break-word;
  transition: border 0.2s ease;
  resize: ${({ resize }) => (resize ? 'initial' : 'none')};

  ${inputStyles};
`;

const Textarea: FunctionComponent<TextareaProps> = props => (
  <StyledTextarea {...props} />
);

export default Textarea;
