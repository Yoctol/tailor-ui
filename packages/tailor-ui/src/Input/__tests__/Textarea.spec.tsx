import React from 'react';

import { render } from 'test/test-utils';

import { Textarea } from '../Textarea';

describe('Textarea', () => {
  it('should render correctly', () => {
    const { container } = render(<Textarea placeholder="placeholder" />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correctly with minRows & maxRows', () => {
    const { container } = render(
      <Textarea id="demo" placeholder="placeholder" rows={3} maxRows={6} />
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
