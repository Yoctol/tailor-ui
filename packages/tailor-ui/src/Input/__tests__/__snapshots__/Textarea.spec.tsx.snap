// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea should render correctly 1`] = `
.c0 {
  word-wrap: break-word;
  resize: none;
  display: block;
  width: 100%;
  max-width: 100%;
  border: 1px solid;
  border-radius: 0.25rem;
  border-color: #e0e0e0;
  outline: none;
  background: #ffffff;
  background-image: none;
  color: #373737;
  line-height: 1.5;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  height: 2rem;
  padding: 4px 0.5rem;
  font-size: 1rem;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
  height: auto;
  -webkit-transition: border 0.2s ease;
  transition: border 0.2s ease;
}

.c0:hover {
  border-color: #8b8b8b;
}

.c0:focus {
  border-color: #191539;
}

.c0:disabled,
.c0 [disabled] {
  border-color: #8b8b8b;
  opacity: 0.5;
  background-color: #e0e0e0;
  cursor: not-allowed;
}

.c0::-webkit-input-placeholder {
  color: #bfbfbf;
}

.c0::-moz-placeholder {
  color: #bfbfbf;
}

.c0:-ms-input-placeholder {
  color: #bfbfbf;
}

.c0::placeholder {
  color: #bfbfbf;
}

<textarea
  class="c0"
  placeholder="placeholder"
  rows="1"
  style="overflow-x: hidden; word-wrap: break-word; overflow-y: hidden;"
/>
`;

exports[`Textarea should render correctly with minRows & maxRows 1`] = `
.c0 {
  word-wrap: break-word;
  resize: none;
  display: block;
  width: 100%;
  max-width: 100%;
  border: 1px solid;
  border-radius: 0.25rem;
  border-color: #e0e0e0;
  outline: none;
  background: #ffffff;
  background-image: none;
  color: #373737;
  line-height: 1.5;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  height: 2rem;
  padding: 4px 0.5rem;
  font-size: 1rem;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
  height: auto;
  -webkit-transition: border 0.2s ease;
  transition: border 0.2s ease;
}

.c0:hover {
  border-color: #8b8b8b;
}

.c0:focus {
  border-color: #191539;
}

.c0:disabled,
.c0 [disabled] {
  border-color: #8b8b8b;
  opacity: 0.5;
  background-color: #e0e0e0;
  cursor: not-allowed;
}

.c0::-webkit-input-placeholder {
  color: #bfbfbf;
}

.c0::-moz-placeholder {
  color: #bfbfbf;
}

.c0:-ms-input-placeholder {
  color: #bfbfbf;
}

.c0::placeholder {
  color: #bfbfbf;
}

<textarea
  class="c0"
  id="demo"
  minrows="3"
  placeholder="placeholder"
  rows="1"
  style="line-height: 1.5em; max-height: 12px;"
/>
`;
